b2luigi/batch/processes/gbasf2.py

```diff
@@ -0,0 +1,673 @@
+           :linenos:
+
+    Handling failed jobs
+        The gbasf2 input wrapper considers the gbasf2 project as failed if any of
+        the jobs in the project failed and reached the maximum number of retries.
+        It then automatically downloads the logs, so please look into them to see what the reason was.
+        For example, it can be that only certain grid sites were affected, so you might want to exclude them
+        by adding the ``"--banned_site ...`` to ``gbasf2_additional_params``.
+
+        You also always reschedule jobs manually with the ``gb2_job_reschedule`` command or delete them with
+        ``gb2_job_delete`` so that the gbasf2 batch process doesn't know they ever
+        existed. Then run just run your luigi task/script again and it will start monitoring the running project
+        again.
+    """
+
+    # directory of the file in which this class is defined
+    _file_dir = os.path.dirname(os.path.realpath(__file__))
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # gbasf2 project name, must be property/attribute, e.g. a luigi parameter
+        # Setting it via a setting.json file is not supported to make sure users set unique project names
+        self.gbasf2_project_name = get_unique_gbasf2_project_name(self.task)
+
+        # Output file directory of the task to wrap with gbasf2, where we will
+        # store the pickled basf2 path and the created steerinfile to execute
+        # that path.
+        task_file_dir = get_task_file_dir(self.task)
+        os.makedirs(task_file_dir, exist_ok=True)
+        #: file name in which the pickled basf2 path from ``self.task.create_path()`` will be stored
+        self.pickle_file_path = os.path.join(task_file_dir, "serialized_basf2_path.pkl")
+        #: file name for steering file that executes pickled path, which will be send to the grid
+        self.wrapper_file_path = os.path.join(task_file_dir, "steering_file_wrapper.py")
+
+        self.log_file_dir = get_log_file_dir(self.task)
+        os.makedirs(self.log_file_dir, exist_ok=True)
+
+        # get dirac user name
+        proxy_info_str = self.setup_dirac_proxy()
+        self.dirac_user = None
+        for line in proxy_info_str.splitlines():
+            if line.startswith("username"):
+                self.dirac_user = line.split(":", 1)[1].strip()
+
+        #: Maximum number of times that each job in the project can be rescheduled until the project is declared as failed.
+        self.max_retries = get_setting("gbasf2_max_retries", default=0, task=self.task)
+
+        #: Store number of times each job had been rescheduled
+        self.n_retries_by_job = Counter()
+
+        #: Local storage for ``n_retries_by_job`` counter
+        # so that it persists even if luigi process is killed and restarted.
+        # TODO: Maybe a small database (e.g. tindydb/pickledb) would be more appropriate
```

Comment:
This is the only TODO comment left in my code. I use a `collections.Counter()` for storing the number of retries for each jobid and whenever I change the counter, I write all of it to the json file. This overwriting of the whole file seems inefficient to me, as I just want to change one single value in that json. Not sure if I can do that easily with json. One idea would be a minimalistic DB, but that would add a dependency and maybe add code complexity, as I just want a key-value store.